// @rozenite/plugin-bridge/dist/index.js
import { useState as d, useEffect as g } from "react";
const _ = "rozenite", D = () => {
  const e = global.__FUSEBOX_REACT_DEVTOOLS_DISPATCHER__, n = e.BINDING_NAME;
  return global[n] != null ? e.initializeDomain(_) : null;
}, f = () => new Promise((e) => {
  const n = (t) => {
    t.name === _ && (global.__FUSEBOX_REACT_DEVTOOLS_DISPATCHER__.onDomainInitialization.removeEventListener(
      n
    ), setTimeout(() => e(t)));
  };
  global.__FUSEBOX_REACT_DEVTOOLS_DISPATCHER__.onDomainInitialization.addEventListener(
    n
  );
}), v = async () => {
  const e = /* @__PURE__ */ new Set();
  let n = D();
  n || (n = await f());
  const t = () => {
    if (!n)
      throw new Error("Domain not initialized");
    return n;
  }, s = (o) => {
    o.name === _ && (n && e.forEach((r) => {
      n.onMessage.removeEventListener(r);
    }), n = o, e.forEach((r) => {
      o.onMessage.addEventListener(r);
    }));
  };
  return global.__FUSEBOX_REACT_DEVTOOLS_DISPATCHER__.onDomainInitialization.addEventListener(
    s
  ), {
    send: (o) => {
      t().sendMessage(o);
    },
    onMessage(o) {
      return e.add(o), t().onMessage.addEventListener(o), {
        remove: () => {
          e.delete(o), t().onMessage.removeEventListener(o);
        }
      };
    },
    close: () => {
      global.__FUSEBOX_REACT_DEVTOOLS_DISPATCHER__.onDomainInitialization.removeEventListener(
        s
      );
    }
  };
}, m = async () => v(), w = async () => {
  const e = /* @__PURE__ */ new Set(), n = (t) => {
    e.forEach((s) => {
      s(t.data);
    });
  };
  return window.addEventListener("message", n), {
    send: (t) => {
      window.parent.postMessage(
        { type: "rozenite-message", payload: t },
        "*"
      );
    },
    onMessage: (t) => (e.add(t), {
      remove: () => {
        e.delete(t);
      }
    }),
    close: () => {
      e.clear(), window.removeEventListener("message", n);
    }
  };
}, C = async () => "__ROZENITE_PANEL__" in window ? w() : m(), T = (e) => typeof e != "object" || e === null || !("type" in e) || !("payload" in e) || !("pluginId" in e) ? null : e, E = /* @__PURE__ */ new Map(), y = async (e) => {
  const n = await C(), t = /* @__PURE__ */ new Map(), s = async (l) => {
    const i = T(l);
    if (!i || i.pluginId !== e)
      return;
    const a = t.get(i.type);
    a != null && a.forEach((u) => {
      u(i.payload);
    });
  }, c = (l, i) => {
    n.send({
      pluginId: e,
      type: l,
      payload: i
    });
  }, o = n.onMessage(s);
  return {
    send: c,
    onMessage: (l, i) => {
      const a = t.get(l) ?? /* @__PURE__ */ new Set();
      return a.add(i), t.set(l, a), {
        remove: () => {
          a.delete(i);
        }
      };
    },
    close: () => {
      t.clear(), o.remove(), n.close();
    }
  };
}, L = async (e) => {
  const n = E.get(e);
  if (n != null)
    return n;
  const t = y(e);
  E.set(e, t);
  const s = await t;
  return E.set(e, s), s;
}, S = ({
  pluginId: e
}) => {
  const [n, t] = d(null), [s, c] = d(null);
  if (g(() => {
    let o = !0, r = null;
    const l = async () => {
      try {
        r = await L(e), o && t(r);
      } catch (a) {
        console.error("Error setting up client", a), o && c(a);
      }
    }, i = async () => {
      try {
        r != null && r.close();
      } catch {
      }
    };
    return l(), () => {
      o = !1, i();
    };
  }, [e]), s != null)
    throw s;
  return n;
};
export {
  L as getRozeniteDevToolsClient,
  S as useRozeniteDevToolsClient
};
